apt/private/package_index.bzl

```diff
@@ -9,19 +9,35 @@ def _fetch_package_index(rctx, url, dist, comp, arch, integrity):
     r = {"success": False, "integrity": None}
 
     decompression_successful = False
+
     for file_type, tool in file_types.items():
         output = "{}/Packages.{}".format(target_triple, file_type)
-        r = rctx.download(
-            url = "{}/dists/{}/{}/binary-{}/Packages.{}".format(url, dist, comp, arch, file_type),
-            output = output,
-            integrity = integrity,
-            allow_fail = True,
-        )
-        if r.success:
-            re = rctx.execute(tool + [output])
-            if re.return_code == 0:
-                decompression_successful = True
-                break
+
+        # Attempt to pull Packages archive from within the `dists` hierarchy first,
+        # if that fails - assume the repository is of type "flat"
+        try_urls = [
+            "{}/dists/{}/{}/binary-{}/Packages.{}".format(url, dist, comp, arch, file_type),
+            "{}/Packages.{}".format(url, file_type),
```

Comment:
How does flat repositories play with different architectures?  in the manifest format we support specifying architectures matrix which i don't see being applied here. 

I believe we should disallow flat repositories to be combined with non flat repositories to prevent confusion.

Reply:
I don't think disallowing flat repos mixed with "classic" repos is really needed.

You can have flat repos with multiple architectures (e.g. https://cloud.r-project.org/bin/linux/debian/bullseye-cran40/ has `i386` and `amd64` packages). The NVIDIA repos are "a bit weirder" since they are flat and have multiple architectures (e.g. https://developer.download.nvidia.com/compute/cuda/repos/debian12/x86_64/ has `i386` and `amd64` packages) but they also seem to have separate flat repos for "other architectures" (e.g. https://developer.download.nvidia.com/compute/cuda/repos/debian12/sbsa/ has `arm64` packages).

Personally, as I mentioned in https://github.com/GoogleContainerTools/rules_distroless/pull/64#issuecomment-2340956897 if there are specific packages in specific architectures I'd just separate them and use another manifest.

Reply:
Hmm, that makes this even more confusing. How does flat repos interact with arches? for instance if you want to install `i386` and `amd64` of a package from   https://developer.download.nvidia.com/compute/cuda/repos/debian12/x86_64/ ? 

I don't see how this PR allows discovery of arches in flat repositories as of today.

Reply:
@thesayyn ah, maybe you are right and there's more fixes needed. I'll try to test the PR and see where it breaks.

Reply:
@thesayyn OK, I think I got it, the main issue that I found so far (and it's been a very quick look so maybe there's more issues / corner cases) is that in `_parse_package_index` we force `arch` to be the one we are using in the for-loop so that's a problem for `Package` indexes that contain multiple architectures as we will get the package info from a different arch.

This is actually happening in this PR and the proof is in the lock and, easier to see, in the test itself :D [Check my comment in `examples/debian_snapshot/test_linux_amd64.yaml`](https://github.com/GoogleContainerTools/rules_distroless/pull/55/commits/b84a604cd93533afd2c222cca03f4f63a35df85d#diff-1558de2257da2042402f3f76a361d4e7f778d60ca44673dcc0f78d8832bd3666), the new `expectedOutput` for the `r-base-core` package is set as `i386` instead of the correct `amd64` precisely because the current code in this PR is selecting the wrong package architecture:
```
(...)
      - r-base-core/now 4.4.1-1~bullseyecran.0 i386 \[installed,local\]
```

Now, the fix that I would propose here is the following:
```diff
diff --git a/apt/private/package_index.bzl b/apt/private/package_index.bzl
index 34f8202..8d6f922 100644
--- a/apt/private/package_index.bzl
+++ b/apt/private/package_index.bzl
@@ -77,7 +77,8 @@ def _parse_package_index(state, contents, arch, root):
 
         if len(pkg.keys()) != 0:
             pkg["Root"] = root
-            util.set_dict(state.packages, value = pkg, keys = (arch, pkg["Package"], pkg["Version"]))
+            arch_ = arch if pkg["Architecture"] == "all" else pkg["Architecture"]
+            util.set_dict(state.packages, value = pkg, keys = (arch_, pkg["Package"], pkg["Version"]))
             last_key = ""
             pkg = {}
```

IMHO this "is not great" because it's duping the packages that are valid across all architectures N times (as many as architectures we have in the manifest). It's not too bad, it's just some storage space wasted and "some mismatch" between the repo structure that is on disk (the per-arch folders) VS the actual contents of the tar files (this can be seen in the dpk-status in the test, e.g. `tzdata` is also `all` but the folder and the "rules_distroless arch" for the package is "amd64" or whatever other arch we are using). But... I don't see how to easily fix this without major changes (and even then I don't know how I'd do it).